### PR TITLE
7 post /api/articles/article id/comments

### PR DESCRIPTION
**Pull request by plaintree**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -212,3 +212,103 @@ describe("6. GET /api/articles/:article_id/comments", () => {
...
+            comment_id: expect.any(Number),
+            body: expect.any(String),
+            article_id: 1,
+            author: expect.any(String),
+            votes: expect.any(Number),
+            created_at: expect.any(String),
+          })
+        );
+      });
+  });
```
Comment (dav3rid) [inline __tests__/app.test.js:239]: As you know the exact details of the comment you are posting, you should be able to assert for specific properties on the response body.
`expect.objectContaining()` is still  valid though to allow for more properties in the future

File: __tests__/app.test.js [hunk 2]
```diff
@@ -212,3 +212,103 @@ describe("6. GET /api/articles/:article_id/comments", () => {
...
+
+  it("status:404, should responds with error message when user does not exists in database but still post comment", () => {
+    const newPost = {
+      username: "some_random_dude",
```
Comment (dav3rid) [inline __tests__/app.test.js:243]: nice test name haha

File: __tests__/app.test.js [hunk 3]
```diff
@@ -212,3 +212,103 @@ describe("6. GET /api/articles/:article_id/comments", () => {
...
+            article_id: 1,
+            author: expect.any(String),
+            votes: expect.any(Number),
+            created_at: expect.any(String),
+          })
+        );
+      });
+  });
+
+  it("status:404, should responds with error message when user does not exists in database but still post comment", () => {
```
Comment (dav3rid) [inline __tests__/app.test.js:241]: Can I check what you mean by "still post comment"?

File: __tests__/app.test.js [hunk 4]
```diff
@@ -212,3 +212,103 @@ describe("6. GET /api/articles/:article_id/comments", () => {
...
+    const newPost = {
+      username: "butter_bridge",
+    };
+    return request(app)
+      .post("/api/articles/1/comments")
+      .send(newPost)
+      .expect(400)
+      .then(({ body }) => {
+        const { msg } = body;
+        expect(msg).toBe("Not Null Violation");
```
Comment (dav3rid) [inline __tests__/app.test.js:266]: I would recommend your error message is something more generic like "bad request" or "malformed request body" - an error message like this begins to expose your data structure eg. using SQL and could invite malicious attempts to mess with it

File: __tests__/app.test.js [hunk 5]
```diff
@@ -212,3 +212,103 @@ describe("6. GET /api/articles/:article_id/comments", () => {
...
+    return request(app)
+      .post("/api/articles/1234523423432423/comments")
+      .send(newPost)
+      .expect(400)
+      .then(({ body }) => {
+        const { msg } = body;
+        expect(msg).toBe("Out Of Range For Type Integer");
+      });
+  });
+});
```
Comment (dav3rid) [inline __tests__/app.test.js:314]: Solid test coverage

File: controllers/articles.controller.js [hunk 6]
```diff
@@ -37,3 +38,20 @@ exports.getArticleComments = (req, res, next) => {
...
+exports.postArticleComment = (req, res, next) => {
+  const {
+    body,
+    params: { article_id },
+  } = req;
+  Promise.all([
+    articlesModel.checkArticleExists(article_id),
+    usersModel.checkUserExists(body.username),
+    articlesModel.postArticleCommentModel(body, article_id),
+  ])
```
Comment (dav3rid) [inline controllers/articles.controller.js:50]: You don't really need to check if the user or article exist here - as they are FOREIGN KEYS in the comments table, if they don't exist, the db.query promise will be rejected for you.

File: models/articles.model.js [hunk 7]
```diff
@@ -32,10 +32,22 @@ exports.getArticleByIdModel = (article_id) => {
 };
 
 exports.getArticleCommentsModel = (article_id) => {
-  SQL = `SELECT comment_id, body, author, votes, created_at
+  let SQL = `SELECT comment_id, body, author, votes, created_at
```
Comment (dav3rid) [inline models/articles.model.js:35]: I know this is for a previous ticket - but well done! Global variables should be avoided, especially now that you're working on a server

File: models/articles.model.js [hunk 8]
```diff
@@ -32,10 +32,22 @@ exports.getArticleByIdModel = (article_id) => {
...
       FROM comments WHERE article_id = $1 ORDER BY created_at DESC;`;
   return db
     .query(SQL, [article_id])
 
     .then(({ rows }) => rows);
 };
+
+exports.postArticleCommentModel = (newComment, article_id) => {
+  const { username, body } = newComment;
+  let SQL = `INSERT INTO comments (body,article_id, author)
```
Comment (dav3rid) [inline models/articles.model.js:46]: could be a `const` if you never reassign

File: models/users.model.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  return db.query(SQL).then(({ rows }) => {
+    const usernameArray = rows.map((user) => user.username);
+    if (!usernameArray.includes(username)) {
+      return Promise.reject({
+        status: 404,
+        msg: "User Not Found",
+      });
+    } else return Promise.resolve(rows);
+  });
+};
```
Comment (dav3rid) [inline models/users.model.js:14]: This function could still be useful later on, but not at this stage

File: models/users.model.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+const db = require("../db/connection");
+
+exports.checkUserExists = (username) => {
+  let SQL = `SELECT username FROM users;`;
+  return db.query(SQL).then(({ rows }) => {
+    const usernameArray = rows.map((user) => user.username);
```
Comment (dav3rid) [inline models/users.model.js:6]: You could take a look at the `some` array method to avoid needing to map here

File: models/users.model.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    if (!usernameArray.includes(username)) {
+      return Promise.reject({
+        status: 404,
+        msg: "User Not Found",
+      });
+    } else return Promise.resolve(rows);
```
Comment (dav3rid) [inline models/users.model.js:12]: No need to explicitly resolve a promise here. A `.then` implicitly returns a promise that resolves to whatever you return form the callback

**Review by dav3rid — APPROVED** [review]

File: __tests__/app.test.js (hunk 3 above)
Comment (plaintree) [reply __tests__/app.test.js:241]: My mistake, I should add in 'still trying to post comment' instead

**Pull request merged**